usr/share/rear/layout/save/GNU/Linux/270_hpraid_layout.sh

```diff
@@ -12,7 +12,7 @@ fi
 # Add $HPSSACLI to the rescue image
 PROGS=( "${PROGS[@]}" $HPSSACLI )
 eval $(grep ON_DIR= $(get_path $HPSSACLI))
-COPY_AS_IS=( "${COPY_AS_IS[@]}" "$HPACUCLI_BIN_INSTALLATION_DIR"  "$HPSSACLI_BIN_INSTALLATION_DIR" )
+COPY_AS_IS=( "${COPY_AS_IS[@]}" "$HPACUCLI_BIN_INSTALLATION_DIR" "$HPSSACLI_BIN_INSTALLATION_DIR" "$SSACLI_BIN_INSTALLATION_DIR")
```

Comment:
As far as I see this cannot work because SSACLI_BIN_INSTALLATION_DIR is nowhere set.

Reply:
Hello, Johannes.

The code is working as such and has been adjusted by @gdha a while ago in another pull request already This is the section of ReaR's debug log:
[...]
++ PROGS=("${PROGS[@]}" $HPSSACLI)
++++ get_path ssacli
++++ type -P ssacli
+++ grep ON_DIR= /sbin/ssacli
++ eval export SSACLI_BIN_INSTALLATION_DIR=/opt/smartstorageadmin/ssacli/bin/
+++ export SSACLI_BIN_INSTALLATION_DIR=/opt/smartstorageadmin/ssacli/bin/
+++ SSACLI_BIN_INSTALLATION_DIR=/opt/smartstorageadmin/ssacli/bin/
++ COPY_AS_IS=("${COPY_AS_IS[@]}" "$HPACUCLI_BIN_INSTALLATION_DIR" "$HPSSACLI_BIN_INSTALLATION_DIR" "$SSACLI_BIN_INSTALLATION_DIR")
+++ get_version ssacli version
+++ TERM=dumb
+++ ssacli version
+++ sed -rn 's/^[^0-9\.]*([0-9]+\.[-0-9a-z\.]+).*$/\1/p'
+++ head -1
++ HPSSACLI_VERSION=3.10.3.0
[...]

Regards,
Holger.

Reply:
@gdha 
can you explain how the variables
HPACUCLI_BIN_INSTALLATION_DIR
HPSSACLI_BIN_INSTALLATION_DIR
SSACLI_BIN_INSTALLATION_DIR
work because from plain looking at the code
I cannot find out how they are set.
For me it looks as if they are always empty.

Reply:
@jsmeix and @gdha 

I try to interpret the lines of code although not beeing a shell expert:
1) Prerequisit: `/sbin/ssacli` is a plain shell script.
2) The `/sbin/ssacli` contains a command
`export SSACLI_BIN_INSTALLATION_DIR=/opt/smartstorageadmin/ssacli/bin/`
3) This string is searched for with `grep ON_DIR= `, executed/set with `eval` and then used in the COPY_AS_IS section for further processing.

IMHO this is not transparent coding without knowing the `ssacli` (or all other names in use by HP) but was implemented as such.
Since HP seems to be quiet creative in renaming this tool I guess a future-proof coding approach would prevent a steady change in this section of code.

Reply:
@hpannenb 
many thanks for your analysis and your explanation.
It helps a lot to understand how that code actually works.

The obscure `eval $(grep ON_DIR= ...` line is initially from
https://github.com/rear/rear/commit/8125ec3b0aef92dedfeb25e330644119fac40f64

For now I would like to not overload this pull request with
an overhaul of that code but an explanatory comment
how that code actually works should be added,
basically exactly what you wrote in
https://github.com/rear/rear/pull/1759#discussion_r175835287
as comment directly in the code in
usr/share/rear/layout/save/GNU/Linux/270_hpraid_layout.sh

I think I cannot overhaul that code myself because I do not have
that HP software.